manual/src/tutorials/parallelism.etex

```diff
@@ -0,0 +1,695 @@
+
+Firstly, immutable values may be freely shared between multiple domains and may
+be accessed in parallel. For mutable data structures such as reference cells,
+arrays and mutable record fields, programmers should avoid \emph{data races}.
+Reference cells, arrays and mutable record fields are said to be
+\emph{non-atomic} data structures. A data race is said to occur when two
+domains concurrently access a non-atomic memory location without
+\emph{synchronisation} and at least one of the accesses is a write. OCaml
+provides a number of ways to introduce synchronisation including atomic
+variables (section \ref{s:par_atomics}) and mutexes (section \ref{s:par_sync}).
+
+Importantly, \textbf{for data race free (DRF) programs, OCaml provides
+sequentially consistent (SC) semantics} -- the observed behaviour of such
+programs can be explained by the interleaving of operations from different
+domains. This property is known as DRF-SC guarantee. Moreover, in OCaml, DRF-SC
+guarantee is modular -- if a part of a program is data race free, then the
+OCaml memory model ensures that those parts have sequential consistency despite
+other parts of the program having data races. Even for programs with data
+races, OCaml provides strong guarantees. While the user may observe non
+sequentially consistent behaviours, there are no crashes.
+
+For more details on the relaxed behaviours in the presence of data races,
+please have a look at the chapter on the hard bits of the memory model
+(chapter \ref{c:memorymodel}).
+
+\section{s:par_atomics}{Atomics}
+
+One way to avoid data races is to use \stdmoduleref{Atomic} variables for
+inter-domain synchronisation. Importantly, unsynchronised concurrent accesses
+to a shared atomic location is not considered a data race. Hence, the program
+will only exhibit sequentially consistent behaviour.
+
+The following program increments a non-atomic counter and an atomic counter in
+parallel.
+
+\begin{caml_example*}{verbatim}
+(* incr.ml *)
+let twice_in_parallel f =
+  let d1 = Domain.spawn f in
+  let d2 = Domain.spawn f in
+  ignore @@ Domain.join d1;
+  ignore @@ Domain.join d2
+
+let plain_ref n =
+  let r = ref 0 in
+  let f () = for _i=1 to n do incr r done in
+  twice_in_parallel f;
+  Printf.printf "Non-atomic ref count: %d\n" !r
+
+let atomic_ref n =
+  let r = Atomic.make 0 in
+  let f () = for _i=1 to n do Atomic.incr r done in
+  twice_in_parallel f;
+  Printf.printf "Atomic ref count: %d\n" (Atomic.get r)
+
+let main () =
+  let n = try int_of_string Sys.argv.(1) with _ -> 1 in
+  plain_ref n;
+  atomic_ref n
+
+let _ = main ()
+\end{caml_example*}
+
+\begin{verbatim}
+$ ocamlopt -o incr.exe incr.ml
+$ ./incr.exe 1_000_000
+Non-atomic ref count: 1187193
+Atomic ref count: 2000000
+\end{verbatim}
+
+As one can see, the final value of the non-atomic counter is lower than one
+might naively expect. In fact, this program exhibits a data race: there are
+unsynchronised concurrent accesses to the non-atomic counter.
```

Comment:
Hi @kayceesrk. There is a difficulty here, as the output of the non-atomic counter is valid SC output. Hence, the example does  does not really illustrate the sentence just before:
> _One way to avoid data races is to use `Atomic` variables for inter-domain synchronisation. Importantly, unsynchronised concurrent accesses to a shared atomic location is not considered a data race. Hence, the program will only exhibit sequentially consistent behaviour._

In fact, the example more illustrates atomicity than SC. May I suggest introducing atomicity before the example. Then present the memory model properties of atomics. This would more or less amount to moving the quoted sentence downwards.

Reply:
You are right. I've modified the text to talk about atomicity instead of data races.